myia/abstract/utils.py

```diff
@@ -745,122 +753,128 @@ def amerge(x1: dtype.TypeMeta, x2, loop, forced, bp):
     return x1 if forced or not changes else type(x1)(rval)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractScalar, x2, loop, forced, bp):
-    values = amerge(x1.values, x2.values, loop, forced, bp)
+def amerge(self, x1: AbstractScalar, x2, forced, bp):
+    values = self(x1.values, x2.values, forced, bp)
     if forced or values is x1.values:
         return x1
     return AbstractScalar(values)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractFunction, x2, loop, forced, bp):
-    values = amerge(x1.get_sync(), x2.get_sync(), loop, forced, bp)
+def amerge(self, x1: AbstractFunction, x2, forced, bp):
+    values = self(x1.get_sync(), x2.get_sync(), forced, bp)
     if forced or values is x1.values:
         return x1
     return AbstractFunction(*values)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractTuple, x2, loop, forced, bp):
+def amerge(self, x1: AbstractTuple, x2, forced, bp):
     args1 = (x1.elements, x1.values)
     args2 = (x2.elements, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
+    merged = self(args1, args2, forced, bp)
     if forced or merged is args1:
         return x1
     return AbstractTuple(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractArray, x2, loop, forced, bp):
+def amerge(self, x1: AbstractArray, x2, forced, bp):
     args1 = (x1.element, x1.values)
     args2 = (x2.element, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
+    merged = self(args1, args2, forced, bp)
     if forced or merged is args1:
         return x1
     return AbstractArray(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractList, x2, loop, forced, bp):
+def amerge(self, x1: AbstractList, x2, forced, bp):
     args1 = (x1.element, x1.values)
     args2 = (x2.element, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
+    merged = self(args1, args2, forced, bp)
     if forced or merged is args1:
         return x1
     return AbstractList(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractClassBase, x2, loop, forced, bp):
+def amerge(self, x1: AbstractClassBase, x2, forced, bp):
     args1 = (x1.tag, x1.attributes, x1.methods, x1.values)
     args2 = (x2.tag, x2.attributes, x2.methods, x2.values)
-    merged = amerge(args1, args2, loop, forced, bp)
-    if forced or merged is args1:
+    merged = self(args1, args2, forced, bp)
+    if forced or merged is args1:  # pragma: no cover
+        # This used to be covered before the `x1 is x2` check in amerge
         return x1
     return type(x1)(*merged)
 
 
 @overload  # noqa: F811
-def amerge(x1: AbstractJTagged, x2, loop, forced, bp):
+def amerge(self, x1: AbstractJTagged, x2, forced, bp):
     args1 = x1.element
     args2 = x2.element
-    merged = amerge(args1, args2, loop, forced, bp)
-    if forced or merged is args1:
+    merged = self(args1, args2, forced, bp)
+    if forced or merged is args1:  # pragma: no cover
+        # This used to be covered before the `x1 is x2` check in amerge
```

Comment:
I'm sure you can cover this with `amerge(AbstractJTagged(1), AbstractJTagged(ANYTHING), forced=True)`